crypto/statetrie/README.md

```diff
@@ -0,0 +1,298 @@
+
+****State trie****
+
+The state trie, commonly known as a prefix tree, is a tree-like data structure
+used for storing an associative array where the keys are sequences of 4-bit
+bytes and the values are SHA-512/256 hashes of the key values.  A proof can be
+provided to the user to show membership of a key value by showing its hash
+provides the necessary missing value to hash to the known root hash.
+
+The trie operates on 'nibbles', which are sequences of 4 bits. This allows for
+a more compact representation compared to standard binary tries and smaller
+proofs.
+
+This trie has built-in support for backing stores, which are essential for
+persistent data storage. It is designed to work seamlessly with both in-memory
+and disk-based storage solutions.
+
+***Key Features:***
+
+Hashing: The trie provides a SHA-512/256 checksum at its root, ensuring data
+integrity.
+
+Adding and removing key/value pairs: Through specific operations, users can
+efficiently add new key-value pairs to the trie or remove existing ones. The
+trie ensures consistent state transitions and optimal space usage during these
+operations.
+
+Child and Merge Operations: The trie supports operations to manage child tries,
+enabling the creation, discard, and merge actions for subtries.
+
+Backstore commit: The trie supports committing changes to the trie to a backing
+store that fuctions like a batched kv interface.
+
+Preloading: Though the trie is designed to keep only parts of it in memory for
+efficiency, it offers a preloading feature to sweep all nodes out of the
+backstore and into memory if required.
+
+***Trie operation and usage:***
+
+Tries are initialized against a backing store (a memory one will be constructed
+if not provided by the user).
+
+```
+mt := MakeTrie(nil)
+key1 := nibbles{0x08, 0x0e, 0x02, 0x08}
+val1 := nibbles{0x03, 0x09, 0x0a, 0x0c}
+key2 := nibbles{0x08, 0x0d, 0x02, 0x08}
+val2 := nibbles{0x03, 0x09, 0x0a, 0x0c}
+
+mt.Add(key1, val1)
+fmt.Println("K1:V1 Hash:", mt.Hash())
+
+mt.Add(key2, val2)
+fmt.Println("K1:V1,K2:V2 Hash:", mt.Hash())
+
+mt.Delete(key2)
+fmt.Println("K1:V1 Hash:", mt.Hash())
+
+mt.Commit()
+```
+
+***Trie nodes:***
+
+There are six possible trie nodes.  Any of the node types can be the root of a
+trie.  Only the first three are committed to the backing store. 
+
+**Leaf nodes**
+
+This value-holding nodes contain the remainder of the search key (the `keyEnd`)
+and the hash of the value.
+
+**Branch nodes**
+
+Branch nodes hold references to 16 children nodes indexed by the next nibble 
+of the search key, plut a "value slot" to hold values for keys th
+at terminate at the branch node. 
+
+**Extension nodes**
+
+Extension nodes contain an addition run of commonly shared key nibbles that
+send you along to the next node.  No value is held at an extension node.
+
+**Parent nodes**
```

Comment:
this term confuses me a bit, because you're using the term "parent node" in two ways.

Reply:
i see what you mean.  maybe I should just call them cow nodes?